### PR TITLE
Investment flow - Part One, Ether Token Investment

### PR DESCRIPTION
**Pull request by trival**
https://github.com/Neufund/platform-frontend/issues/1212

### CODE REVIEW

File: app/components/modals/txSender/TxSender.tsx [hunk 1]
```diff
@@ -81,7 +100,7 @@ function renderBody({
       return <WithdrawSuccess />;
 
     case "ERROR_SIGN":
-      return <div>Error occured!</div>;
+      return <div>Error occured!!!!</div>;
```
Comment (Moejoe90) [inline app/components/modals/txSender/TxSender.tsx]: #https://github.com/Neufund/platform-frontend/issues/1299 related

File: app/components/modals/txSender/TxSender.tsx [hunk 2]
```diff
@@ -25,48 +28,64 @@ interface IStateProps {
...
+}
+export type TSummaryComponentProps = ITxSummaryStateProps & ITxSummaryDispatchProps;
+
+function getInitComponent(type: TxSenderType): JSX.Element {
+  switch (type) {
+    case "INVEST":
+      return <InvestmentSelection />;
+    case "WITHDRAW":
+      return <Withdraw />;
+  }
```
Comment (Moejoe90) [inline app/components/modals/txSender/TxSender.tsx]: can we add a default state with a throw? for debugging purpose

File: app/components/modals/txSender/TxSender.tsx [hunk 3]
```diff
@@ -25,48 +28,64 @@ interface IStateProps {
...
+}
+
+function getSummaryComponent(type: TxSenderType): JSX.Element {
+  switch (type) {
+    case "INVEST":
+      return <InvestmentSummary />;
+    case "WITHDRAW":
+      return <WithdrawSummary />;
```
Comment (Moejoe90) [inline app/components/modals/txSender/TxSender.tsx]: also here

File: app/components/modals/txSender/investment-flow/Investment.tsx [hunk 4]
```diff
@@ -1,143 +1,308 @@
...
+      return {
+        etherPriceEur: selectEtherPriceEur(state.tokenPrice),
+        euroValue: state.investmentFlow.euroValueUlps,
+        ethValue: state.investmentFlow.ethValueUlps,
+        errorState: state.investmentFlow.errorState,
+        gasCostEth: selectInvestmentGasCostEth(state.investmentFlow),
+        investmentType: state.investmentFlow.investmentType,
+        wallets: createWallets(state),
+        calculatedContribution: state.investmentFlow.calculatedContribution,
+        minTicketEur: (state.investmentFlow.eto && state.investmentFlow.eto.minTicketEur!) || 0,
```
Comment (Moejoe90) [inline app/components/modals/txSender/investment-flow/Investment.tsx]: can we move this to a selector

File: app/components/modals/txSender/investment-flow/Summary.tsx [hunk 5]
```diff
@@ -97,12 +119,34 @@ export const InvestmentSummary = injectIntl(
...
+    return {
+      agreementUrl: "fufu",
+      companyName: eto.company.name!,
+      etoAddress: eto.etoId,
+      investmentEth: i.ethValueUlps,
+      investmentEur: i.euroValueUlps,
+      gasCostEth: selectInvestmentGasCostEth(i),
+      equityTokens: i.calculatedContribution!.equityTokenInt.toString(),
+      estimatedReward: i.calculatedContribution!.neuRewardUlps.toString(),
+      etherPriceEur: state.tokenPrice.tokenPriceData!.etherPriceEur,
```
Comment (Moejoe90) [inline app/components/modals/txSender/investment-flow/Summary.tsx]: can we change to selectors

File: app/components/shared/Money.tsx [hunk 6]
```diff
@@ -1,8 +1,8 @@
 import * as cn from "classnames";
 import * as React from "react";
 import { MONEY_DECIMALS } from "../../config/constants";
+import { formatMoney, formatThousands } from "../../utils/Money.utils";
 import * as styles from "./Money.module.scss";
```
Comment (Moejoe90) [inline app/components/shared/Money.tsx]: import fix

File: app/lib/web3/ContractsService.ts [hunk 7]
```diff
@@ -88,6 +90,8 @@ export class ContractsService {
     this.euroToken = await create(EuroToken, web3, euroTokenAddress);
     this.etherToken = await create(EtherToken, web3, etherTokenAddress);
 
+    // this.etherToken.rawWeb3Contract.transfer.getData()
```
Comment (Moejoe90) [inline app/lib/web3/ContractsService.ts]: remove comment

File: app/components/dashboard/eto-list/EtoList.tsx [hunk 8]
```diff
@@ -29,8 +29,10 @@ export const EtoListComponent: React.SFC<IProps> = ({ etos, startInvestmentFlow
       <Col xs={12} lg={6} key={index}>
         {/* TODO: implement real ETO Card here */}
         <div className="mb-3">
-          <h4>{eto.company.brandName}</h4>
-          <Button onClick={() => startInvestmentFlow(eto)}>invest now</Button>
+          <h4>{`${eto.company.brandName} - ${eto.company.name}`}</h4>
+          {eto.state === "on_chain" && (
+            <Button onClick={() => startInvestmentFlow(eto)}>invest now</Button>
```
Comment (desfero) [inline app/components/dashboard/eto-list/EtoList.tsx]: @trival do we need to add to translations 'invest now'?

File: app/components/dashboard/eto-list/EtoList.tsx [hunk 9]
```diff
@@ -29,8 +29,10 @@ export const EtoListComponent: React.SFC<IProps> = ({ etos, startInvestmentFlow
       <Col xs={12} lg={6} key={index}>
         {/* TODO: implement real ETO Card here */}
```
Comment (desfero) [inline app/components/dashboard/eto-list/EtoList.tsx]: Do we still need TODO comment?

File: app/components/modals/txSender/TxSender.tsx [hunk 10]
```diff
@@ -25,48 +28,68 @@ interface IStateProps {
...
+function isBigModal(props: Props): boolean {
+  return props.state === "INIT" && props.type === "INVEST";
+}
+
 export const TxSenderModalComponent: React.SFC<Props> = props => {
   const { isOpen, onCancel } = props;
 
   return (
-    <Modal isOpen={isOpen} toggle={onCancel}>
+    <Modal isOpen={isOpen} toggle={onCancel} className={cn(isBigModal(props) && "big")}>
```
Comment (desfero) [inline app/components/modals/txSender/TxSender.tsx]: From the readability reason, what you think about
`className={cn({ big: isBigModal(props)} )`?

File: app/components/modals/txSender/TxSender.tsx [hunk 11]
```diff
@@ -25,48 +28,68 @@ interface IStateProps {
...
+export interface ITxSummaryStateProps {
+  txData: ITxData;
+}
+export interface ITxSummaryDispatchProps {
   onAccept: () => any;
-  data: ITxData;
+}
+export type TSummaryComponentProps = ITxSummaryStateProps & ITxSummaryDispatchProps;
+
+function getInitComponent(type?: TxSenderType): JSX.Element {
```
Comment (desfero) [inline app/components/modals/txSender/TxSender.tsx]: From the technical point of view you always return already processed jsx element (React.createElement...) so maybe it's worth to rename the method to 'getInitElement'?

File: app/components/modals/txSender/TxSender.tsx [hunk 12]
```diff
@@ -25,48 +28,68 @@ interface IStateProps {
...
+    case "INVEST":
+      return <InvestmentSelection />;
+    case "WITHDRAW":
+      return <Withdraw />;
+    default:
+      throw new Error("Transaction type needs to be set at transaction init state");
+  }
+}
+
+function getSummaryComponent(type?: TxSenderType): JSX.Element {
```
Comment (desfero) [inline app/components/modals/txSender/TxSender.tsx]: From the technical point of view you always return already processed jsx element (React.createElement...) so maybe it's worth to rename the method to 'getInitElement'?

File: app/components/modals/txSender/TxSender.tsx (hunk 11 above)
Comment (desfero) [reply app/components/modals/txSender/TxSender.tsx]: Or even, let's do it react way. You can simply convert that function to stateless component `TxSenderInitComponent`.

File: app/components/modals/txSender/investment-flow/Summary.tsx [hunk 13]
```diff
@@ -1,53 +1,84 @@
...
 import { DocumentLink } from "../../../shared/DocumentLink";
 import { Heading } from "../../../shared/modals/Heading";
 import { InfoList } from "../shared/InfoList";
 import { InfoRow } from "../shared/InfoRow";
+import { ITxSummaryDispatchProps } from "../TxSender";
 
 import * as neuIcon from "../../../../assets/img/neu_icon.svg";
 import * as tokenIcon from "../../../../assets/img/token_icon.svg";
+import { MONEY_DECIMALS } from "../../../../config/constants";
+import { selectEtherPriceEur } from "../../../../modules/shared/tokenPrice/selectors";
```
Comment (desfero) [inline app/components/modals/txSender/investment-flow/Summary.tsx]: fix imports

**Comment by desfero** [comment]
@SergejSoft this one for sure please merge with squash ;)

File: app/components/dashboard/eto-list/EtoList.tsx (hunk 8 above)
Comment (trival) [reply app/components/dashboard/eto-list/EtoList.tsx]: this is just a dummy component, the real invest button will be on eto details widget

File: app/components/dashboard/eto-list/EtoList.tsx (hunk 9 above)
Comment (trival) [reply app/components/dashboard/eto-list/EtoList.tsx]: yes, as this is only placeholder

File: app/components/modals/txSender/TxSender.tsx (hunk 10 above)
Comment (trival) [reply app/components/modals/txSender/TxSender.tsx]: agree

**Pull request merged**